### PR TITLE
Implementation for non byte aligned metadata and header fields

### PR DESCRIPTION
**Pull request by github-sajan**

### CODE REVIEW

File: backends/dpdk/backend.cpp [hunk 1]
```diff
@@ -52,6 +53,13 @@ void DpdkBackend::convert(const IR::ToplevelBlock *tlb) {
         new P4::ClearTypeMap(typeMap),
         new P4::TypeChecking(refMap, typeMap),
         // TBD: implement dpdk lowering passes instead of reusing bmv2's lowering pass.
+
```
Comment (usha1830) [inline backends/dpdk/backend.cpp:55]: Please group these passes together, so that we accidentally don't change the sequence

File: backends/dpdk/backend.cpp [hunk 2]
```diff
@@ -52,6 +53,13 @@ void DpdkBackend::convert(const IR::ToplevelBlock *tlb) {
...
+        new AlignHdrMetaField(typeMap, refMap, &structure),
+        new P4::ClearTypeMap(typeMap),
+        new P4::TypeChecking(refMap, typeMap, true),
+        new P4::DoRemoveLeftSlices(typeMap),
+        new P4::ClearTypeMap(typeMap),
```
Comment (usha1830) [inline backends/dpdk/backend.cpp]: Is it not sufficient to have ClearTypeMap and TypeChecking called only once after DoRemoveLeftSlices?

File: backends/dpdk/dpdkArch.cpp [hunk 3]
```diff
@@ -467,6 +467,127 @@ bool CollectMetadataHeaderInfo::preorder(const IR::Type_Struct *s) {
     return true;
 }
 
+const IR::Node *AlignHdrMetaField::preorder(IR::Type_StructLike *st) {
+    auto fields = new IR::IndexedVector<IR::StructField>;
+    if (st->is<IR::Type_Header>()) {
+        auto h = st->to<IR::Type_Header>();
+        unsigned size_sum_so_far = 0;
+        for (auto field : st->fields) {
+            if (auto t = (*field).type->to<IR::Type_Bits>()) {
```
Comment (Sosutha) [inline backends/dpdk/dpdkArch.cpp]: Please add a check if the type->is<IR::Type_Bits>, If there is any other type, this might result in seg fault

File: backends/dpdk/dpdkArch.cpp [hunk 4]
```diff
@@ -467,6 +467,127 @@ bool CollectMetadataHeaderInfo::preorder(const IR::Type_Struct *s) {
...
+                auto width = t->width_bits();
+                if (t->width_bits() % 8 != 0) {
```
Comment (Sosutha) [inline backends/dpdk/dpdkArch.cpp]: variable width can be used

File: backends/dpdk/dpdkArch.cpp [hunk 5]
```diff
@@ -467,6 +467,127 @@ bool CollectMetadataHeaderInfo::preorder(const IR::Type_Struct *s) {
...
+                    size_sum_so_far += t->width_bits();
```
Comment (Sosutha) [inline backends/dpdk/dpdkArch.cpp]: variable width can be used

**Review by Sosutha — CHANGES_REQUESTED** [review]

File: testdata/p4_16_samples_outputs/pna-issue3041.p4.spec [hunk 6]
```diff
@@ -42,10 +40,12 @@ struct a2_arg_t {
 struct main_metadata_t {
 	bit<32> pna_main_input_metadata_input_port
 	bit<32> pna_main_output_metadata_output_port
-	bit<32> MainParserT_parser_tmp
+	bit<8> MainParserT_parser_tmp
```
Comment (usha1830) [inline testdata/p4_16_samples_outputs/pna-issue3041.p4.spec:43]: Why is changed? Is this expected?

**Review by usha1830 — CHANGES_REQUESTED** [review]

**Comment by usha1830** [comment]
Please also add appropriate comments explaining the changes. Also include one example each for all the new passes where you can show the P4 before and after the pass is run.

File: testdata/p4_16_samples_outputs/pna-issue3041.p4.spec (hunk 6 above)
Comment (github-sajan) [reply testdata/p4_16_samples_outputs/pna-issue3041.p4.spec:43]: This is expected. MainParserT_parser_tmp is used for processing version=ihl field since it is a 8-bit field

File: backends/dpdk/backend.cpp (hunk 2 above)
Comment (github-sajan) [reply backends/dpdk/backend.cpp]: No, It is expecting after both  passses

File: backends/dpdk/backend.cpp (hunk 1 above)
Comment (github-sajan) [reply backends/dpdk/backend.cpp:55]: Addressed

File: backends/dpdk/dpdkArch.cpp (hunk 5 above)
Comment (github-sajan) [reply backends/dpdk/dpdkArch.cpp]: Addressed

File: backends/dpdk/dpdkArch.cpp (hunk 4 above)
Comment (github-sajan) [reply backends/dpdk/dpdkArch.cpp]: Addressed

File: backends/dpdk/dpdkArch.cpp (hunk 3 above)
Comment (github-sajan) [reply backends/dpdk/dpdkArch.cpp]: Addressed

File: backends/dpdk/DpdkXfail.cmake [hunk 7]
```diff
@@ -23,6 +23,18 @@ p4c_add_xfail_reason("dpdk"
   testdata/p4_16_samples/psa-action-profile2.p4
   )
 
+p4c_add_xfail_reason("dpdk"
+  "not implemented"
+  testdata/p4_16_samples/psa-example-dpdk-byte-alignment_3.p4
+  testdata/p4_16_samples/psa-example-dpdk-byte-alignment_4.p4
+  testdata/p4_16_samples/psa-example-incremental-checksum.p4
```
Comment (usha1830) [inline backends/dpdk/DpdkXfail.cmake]: This has started failing now? or was it failing with some other error earlier?

**Review by usha1830 — CHANGES_REQUESTED** [review]
Please add code comments and examples snippets as requested earlier.

File: backends/dpdk/DpdkXfail.cmake (hunk 7 above)
Comment (github-sajan) [reply backends/dpdk/DpdkXfail.cmake]: Earlier this was passing. With byte alignment, This will fail. However since this pattern is present in egress control, now this is passing as egress control is removed by default. So i have added a testcase with similar pattern in ingress control

**Review by usha1830 — APPROVED** [review]

File: backends/dpdk/dpdkArch.cpp [hunk 8]
```diff
@@ -467,6 +467,215 @@ bool CollectMetadataHeaderInfo::preorder(const IR::Type_Struct *s) {
...
+            ::error("8-bit Alignment for Header Structure '%1%' is not possible", st->name.name);
+            return st;
+        }
+        return new IR::Type_Header(IR::ID(st->name), st->annotations, *fields);
+    }
+
+    return st;
+}
+
+/* This function process the occurance of non-aligned fields and replaces it with
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.cpp:606]: occurrence*

File: testdata/p4_16_samples_outputs/invalid-hdr-warnings2-first.p4 [hunk 9]
```diff
@@ -86,3 +86,4 @@ control ComputeChecksumI(inout H hdr, inout M meta) {
 }
 
 V1Switch<H, M>(ParserI(), VerifyChecksumI(), IngressI(), EgressI(), ComputeChecksumI(), DeparserI()) main;
+
```
Comment (usha1830) [inline testdata/p4_16_samples_outputs/invalid-hdr-warnings2-first.p4:89]: Please remove these unrelated changes

File: testdata/p4_16_samples_outputs/issue1001-1-bmv2-first.p4 [hunk 10]
```diff
@@ -10,7 +10,7 @@ enum bit<8> PreservedFieldList {
 }
 
 struct Meta {
-    @field_list(PreservedFieldList.Field)
+    @field_list(PreservedFieldList.Field)
```
Comment (usha1830) [inline testdata/p4_16_samples_outputs/issue1001-1-bmv2-first.p4:13]: Here as well, this change is unrelated

**Review by usha1830 — APPROVED** [review]
Looks fine. Please remove unrelated reference output files.

File: testdata/p4_16_samples_outputs/issue1001-1-bmv2-first.p4 (hunk 10 above)
Comment (github-sajan) [reply testdata/p4_16_samples_outputs/issue1001-1-bmv2-first.p4:13]: I will address this while raising review in open source

File: testdata/p4_16_samples_outputs/invalid-hdr-warnings2-first.p4 (hunk 9 above)
Comment (github-sajan) [reply testdata/p4_16_samples_outputs/invalid-hdr-warnings2-first.p4:89]: I will address this while raising review in open source

File: backends/dpdk/dpdkArch.cpp (hunk 8 above)
Comment (github-sajan) [reply backends/dpdk/dpdkArch.cpp:606]: I will address this while raising review in open source

**Review by Sosutha — APPROVED** [review]
